src/contracts/libraries/GPv2Encoding.sol

```diff
@@ -0,0 +1,128 @@
+    /// Orders are tightly packed in order to reduce calldata and associated gas
+    /// costs. They contain the following fields:
+    /// ```
+    /// struct EncodedOrder {
+    ///     uint8 sellTokenIndex;
+    ///     uint8 buyTokenIndex;
+    ///     uint256 sellAmount;
+    ///     uint256 buyAmount;
+    ///     uint32 validTo;
+    ///     uint32 nonce;
+    ///     uint256 tip;
+    ///     uint8 flags;
+    ///     uint256 executedAmount;
+    ///     Signature {
+    ///         uint8 v;
+    ///         bytes32 r;
+    ///         bytes32 s;
+    ///     } signature;
+    /// }
+    /// ```
+    /// @param tokens The list of tokens included in the settlement. The encoded
+    /// token indices part of the order map to tokens in this array.
+    /// @param encodedOrder The order as encoded calldata bytes.
+    /// @param order The memory location to decode the order to.
+    function decodeSignedOrder(
+        bytes32 domainSeparator,
+        IERC20[] calldata tokens,
+        bytes calldata encodedOrder,
+        Order memory order
+    ) internal pure {
+        // NOTE: This is currently unnecessarily gas inefficient. Specifically,
+        // there is a potentially extraneous check to the encoded order length
+        // (this can be verified once for the total encoded orders length).
+        // Additionally, Solidity generates bounds checks for each `abi.decode`
+        // and slice operation. Unfortunately using `assmebly { calldataload }`
+        // is quite ugly here since there is no mechanism to get calldata
+        // offsets (like there is for memory offsets) without manual
+        // computation, which is brittle as changes to the calling function
+        // signature would require manual adjustments to the computation. Once
+        // gas benchmarking is set up, we can evaluate if it is worth the extra
+        // effort.
+
+        require(
+            encodedOrder.length == ORDER_STRIDE,
+            "GPv2: malformed order data"
+        );
+
+        order.sellTokenIndex = uint8(encodedOrder[0]);
+        order.sellToken = tokens[order.sellTokenIndex];
+        order.buyTokenIndex = uint8(encodedOrder[1]);
+        order.buyToken = tokens[order.buyTokenIndex];
+        order.sellAmount = abi.decode(encodedOrder[2:], (uint256));
+        order.buyAmount = abi.decode(encodedOrder[34:], (uint256));
+        order.validTo = uint32(
+            abi.decode(encodedOrder[66:], (uint256)) >> (256 - 32)
+        );
+        order.nonce = uint32(
+            abi.decode(encodedOrder[70:], (uint256)) >> (256 - 32)
+        );
+        order.tip = abi.decode(encodedOrder[74:], (uint256));
+        uint8 flags = uint8(encodedOrder[106]);
+        order.kind = OrderKind((flags >> ORDER_KIND_BIT) & 0x1);
+        order.partiallyFillable =
+            (flags >> ORDER_PARTIALLY_FILLABLE_BIT) & 0x01 == 0x01;
+        order.executedAmount = abi.decode(encodedOrder[107:], (uint256));
```

Comment:
executedAmount shouldn't be part of the digest and the signature right? It's not something that is stored in the orderbook, just extra information for a settlement. And after a partially settled order has been partially settled we can never settle it again anyway.

Reply:
It is not. So, further down when we compute the keccak hash of the order data, (what we use for the `ecrecover`, essentially what is being signed) we do not include the executed amount. It is, however, part of the order data needed for settlement.

Reply:
I was also quite confused at first. Maybe it helps to add a comment where you define the `Order` struct that explains how the order of the elements in the struct is important:
1. first should appear all parameters that are supposed to be signed by the user in an order
2. then (from `executedAmount` on down), you add parameters describing how orders are executed.

The threshold defines the magic number 10 used in the assembly below.

Reply:
If you wanted to make it more obvious you could split the Order struct into two: `SettlementInfo` which contains an `Order` struct that only has the fields for which we check signature. But not sure if this has performance implications otherwise.

Reply:
I had wanted to do that but I ended up just creating a struct with pointers to other in memory structs, instead of being sequencially laid out in memory as I expected. I'll double check if there is some other way to do this.

Reply:
So I double checked and indeed:
```solidity
    struct Foo {
        address a;
        Bar b;
    }
    struct Bar {
        uint256 x;
        uint8 y;
    }
```

Is represented as:
```
0x80: 000000000000000000000000address.................................
0xa0: 00000000000000000000000000000000000000000000000000000000000000c0
0xc0: x...............................................................
0xe0: 00000000000000000000000000000000000000000000000000000000000000y.
```

So, indeed the structs do not appear to get flattened (unfortunately).

I will add more documentation to make this clearer that:
- The order of the fields is important
- The order struct is really an executed order that contains the order data, with additional properties.

Reply:
Added documentation to the order struct